### PR TITLE
feat(quantic): update search bar look

### PR DESCRIPTION
**Pull request by fbeaupreCoveo**
### Related Issue 
[SFINT-3830](https://coveord.atlassian.net/browse/SFINT-3830)

### Proposed Changes
Updated the search bar look based on [these mockups](https://www.figma.com/file/02bJKPyf2kwP6He76B42kb/Search-page-components?node-id=69%3A99)

- With the search button
Screen Shot 2021-07-16 at 9 17 54 AM

- Without the search button
Screen Shot 2021-07-16 at 9 18 37 AM

### CODE REVIEW

File: packages/quantic/typings/lwc/engine.d.ts [hunk 1]
```diff
@@ -126,6 +126,20 @@ declare module 'lwc' {
         compareDocumentPosition(otherNode: Node): number;
         contains(otherNode: Node): boolean;
         elementFromPoint(x: number, y: number): Element | null;
+        querySelector<K extends keyof HTMLElementTagNameMap>(
```
Comment (nathanlb) [inline packages/quantic/typings/lwc/engine.d.ts:129]: calling querySelector without ts-ignore was raising typing errors. Upon investigation I noticed salesforce has updated their LWC typings since we first imported them into Quantic. So I updated our typings from [lwc repo](https://github.com/salesforce/lwc/blob/5100fc0b3db3238ae97316c494d06fe328c6bf1f/packages/lwc/types.d.ts#L129)

Comment (fbeaupreCoveo) [reply packages/quantic/typings/lwc/engine.d.ts:129]: Thank you for the fix!

File: packages/quantic/force-app/main/default/lwc/quanticSearchBox/quanticSearchBox.css [hunk 2]
```diff
@@ -0,0 +1,30 @@
+.searchbox-container{
```
Comment (erocheleau) [inline packages/quantic/force-app/main/default/lwc/quanticSearchBox/quanticSearchBox.css:1]: Do you know if any of these can be achieved with classes from slds?

File: packages/quantic/force-app/main/default/lwc/quanticSearchBox/quanticSearchBox.html [hunk 3]
```diff
@@ -8,21 +8,60 @@
...
-              onkeyup={onKeyup}
-              onfocus={onFocus}
-              onblur={onBlur}
-              aria-autocomplete="list"
-              aria-controls="listbox-id"
-              autocomplete="off"
-              role="textbox"
-            />
-          </div>
+          <template if:true={withoutSubmitButton}>
```
Comment (erocheleau) [inline packages/quantic/force-app/main/default/lwc/quanticSearchBox/quanticSearchBox.html:11]: Seems to be a lot of repeated code between the two different templates, 

Would you prefer to make a `get searchBoxContainerClass()` JS method that would return the correct class string and then adapt your CSS for that?

And only surround the lightning-icon with the template if:true ?

File: packages/quantic/force-app/main/default/lwc/quanticSearchBox/quanticSearchBox.html [hunk 4]
```diff
@@ -40,7 +79,7 @@
                     <div class="slds-media slds-listbox__option slds-listbox__option_plain slds-media_small" role="option">
                       <span class="slds-media__figure slds-listbox__option-icon"></span>
                       <span class="slds-media__body">
-                        <span class="slds-truncate" title={suggestion.value}>{suggestion.value}</span>
+                        <lightning-formatted-rich-text class="slds-truncate" value={suggestion.value}></lightning-formatted-rich-text>
```
Comment (erocheleau) [inline packages/quantic/force-app/main/default/lwc/quanticSearchBox/quanticSearchBox.html:82]: Why this change?

File: packages/quantic/force-app/main/default/lwc/quanticSearchBox/quanticSearchBox.js [hunk 5]
```diff
@@ -35,22 +43,39 @@ export default class QuanticSearchBox extends LightningElement {
   input;
   /** @type {import("lwc").HTMLElementTheGoodPart} */
   combobox;
-
+  /** @type {HTMLButtonElement} */
+  clearButton;
+  /** @type {HTMLElement} */
+  closeButton
```
Comment (erocheleau) [inline packages/quantic/force-app/main/default/lwc/quanticSearchBox/quanticSearchBox.js:49]: What is a closeButton exactly?

File: packages/quantic/force-app/main/default/lwc/quanticSearchBox/quanticSearchBox.js [hunk 6]
```diff
@@ -102,7 +132,7 @@ export default class QuanticSearchBox extends LightningElement {
     this.resetHighlighted();
     suggestion.setAttribute('aria-selected', 'true');
     suggestion.classList.add('slds-has-focus');
-    this.input.value = suggestion.textContent;
+    this.input.value = suggestion.innerText;
```
Comment (erocheleau) [inline packages/quantic/force-app/main/default/lwc/quanticSearchBox/quanticSearchBox.js:135]: This is related to the rich text change above, is it because we want to be able to include html for highlighting?

File: packages/quantic/force-app/main/default/lwc/quanticSearchBox/quanticSearchBox.html (hunk 4 above)
Comment (nathanlb) [reply packages/quantic/force-app/main/default/lwc/quanticSearchBox/quanticSearchBox.html:82]: for highlighted substrings in query suggestions

File: packages/quantic/force-app/main/default/lwc/quanticSearchBox/quanticSearchBox.js (hunk 5 above)
Comment (fbeaupreCoveo) [reply packages/quantic/force-app/main/default/lwc/quanticSearchBox/quanticSearchBox.js:49]: The mockups have two different icons for the `X` button to clear the input. One is called clear and the other is called close, that's why I called one clearButton and the other closeButton. They do the same thing but I couldn't give them the same Id and use only one element for both.

File: packages/quantic/force-app/main/default/lwc/quanticSearchBox/quanticSearchBox.js (hunk 6 above)
Comment (fbeaupreCoveo) [reply packages/quantic/force-app/main/default/lwc/quanticSearchBox/quanticSearchBox.js:135]: It's because with `textContent` there was a bug where if the user navigates through the suggestions with the keyboard, the input in the search box would not update and when the user pressed enter, it would submit a search with an empty string query. Using `innerText` fixed that.

File: packages/quantic/force-app/main/default/lwc/quanticSearchBox/quanticSearchBox.js [hunk 7]
```diff
@@ -35,22 +43,39 @@ export default class QuanticSearchBox extends LightningElement {
   input;
   /** @type {import("lwc").HTMLElementTheGoodPart} */
   combobox;
-
+  /** @type {HTMLButtonElement} */
+  clearButton;
```
Comment (lbergeron) [inline packages/quantic/force-app/main/default/lwc/quanticSearchBox/quanticSearchBox.js:47]: Do you really need both `clearButton` and `closeButton`? You seem to be applying the same changes to both anyways, so it would remove some duplication.

File: packages/quantic/force-app/main/default/lwc/quanticSearchBox/quanticSearchBox.js [hunk 8]
```diff
@@ -61,13 +86,18 @@ export default class QuanticSearchBox extends LightningElement {
...
-      // @ts-ignore
       this.input = this.template.querySelector('input');
     }
     if (!this.combobox) {
-      // @ts-ignore
       this.combobox = this.template.querySelector('.slds-combobox');
     }
+    if (this.withoutSubmitButton && !this.clearButton) {
+      this.clearButton = this.template.querySelector('.slds-button__icon');
+      this.clearButton.hidden = true;
```
Comment (lbergeron) [inline packages/quantic/force-app/main/default/lwc/quanticSearchBox/quanticSearchBox.js:96]: Why are you setting `hidden = true` in this specific case?

File: packages/quantic/force-app/main/default/lwc/quanticSearchBox/quanticSearchBox.css [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+    display: flex;
+    position: relative;
+    border-color: #e5e8e8;
+    height: 32px;
+    border-radius: 4px;
+    box-sizing: border-box;
+    background: #FFFFFF;
+}
+
+.searchbox_input {
```
Comment (lbergeron) [inline packages/quantic/force-app/main/default/lwc/quanticSearchBox/quanticSearchBox.css:11]: Using the BEM naming convention would be nice and more aligned with SLDS. Here it would mean renaming the classes like this:

`searchbox-container` --> `searchbox__container`
`searchbox_input` --> `searchbox__input`
`search-button` --> `searchbox__search-button` or even `searchbox__submit-button` to match the exposed attribute name.

File: packages/quantic/force-app/main/default/lwc/quanticSearchBox/quanticSearchBox.js (hunk 5 above)
Comment (nathanlb) [reply packages/quantic/force-app/main/default/lwc/quanticSearchBox/quanticSearchBox.js:49]: They should never both appear at the same time, can't it just be one reference to `clearButton`, the only thing that changes about it is the name of the icon used right?

File: packages/quantic/force-app/main/default/lwc/quanticSearchBox/quanticSearchBox.js [hunk 10]
```diff
@@ -14,17 +17,22 @@ export default class QuanticSearchBox extends LightningElement {
...
+    }));
   }
-
   /** @type {string} */
   @api engineId;
   /** @type {string} */
-  @api placeholder = 'Search';
+  @api placeholder = 'Search...';
+   /** @type {boolean} */
+   @api withoutSubmitButton = false;
```
[the remark this answers is not in the document]
Comment (fbeaupreCoveo) [reply packages/quantic/force-app/main/default/lwc/quanticSearchBox/quanticSearchBox.js:32]: I initially had it in the affirmative form but I got the following error: `Boolean public properties should not be initialized to "true", consider initializing the property to "false".` I can disable the error check for this line though, should I?

File: packages/quantic/force-app/main/default/lwc/quanticSearchBox/quanticSearchBox.js (hunk 8 above)
Comment (fbeaupreCoveo) [reply packages/quantic/force-app/main/default/lwc/quanticSearchBox/quanticSearchBox.js:96]: forgot to remove it after I started using `slds-hidden` instead. I'll remove it now

File: packages/quantic/force-app/main/default/lwc/quanticSearchBox/quanticSearchBox.js (hunk 5 above)
Comment (fbeaupreCoveo) [reply packages/quantic/force-app/main/default/lwc/quanticSearchBox/quanticSearchBox.js:49]: Actually pay no attention to my argument ^, it was nonsense. It works using `clearButton` for both elements

File: packages/quantic/force-app/main/default/lwc/quanticSearchBox/quanticSearchBox.js (hunk 7 above)
Comment (fbeaupreCoveo) [reply packages/quantic/force-app/main/default/lwc/quanticSearchBox/quanticSearchBox.js:47]: You're right it works using only `clearButton`. For some reason I thought i needed to use two elements. Thanks!

File: packages/quantic/force-app/main/default/lwc/quanticSearchBox/quanticSearchBox.js (hunk 10 above)
[the remark this answers is not in the document]
Comment (lbergeron) [reply packages/quantic/force-app/main/default/lwc/quanticSearchBox/quanticSearchBox.js:32]: Good to know. I wasn't aware of that.
[This doc](https://developer.salesforce.com/docs/component-library/documentation/en/lwc/js_props_boolean) explains why it works this way.

File: packages/quantic/force-app/main/default/lwc/quanticSearchBox/quanticSearchBox.html (hunk 3 above)
Comment (fbeaupreCoveo) [reply packages/quantic/force-app/main/default/lwc/quanticSearchBox/quanticSearchBox.html:11]: You're right that would probably better. I think I would need another template if:false around the left lightning-icon, and to add a class to the input for the add-on after checking `withoutSubmitButton`. I'll give it a go, thanks!

File: packages/quantic/force-app/main/default/lwc/quanticSearchBox/quanticSearchBox.js [hunk 11]
```diff
@@ -14,17 +17,25 @@ export default class QuanticSearchBox extends LightningElement {
...
+    }));
   }
-
   /** @type {string} */
   @api engineId;
   /** @type {string} */
-  @api placeholder = 'Search';
+  @api placeholder = 'Search...';
+   /** @type {boolean} */
+   withoutSubmitButton = false;
```
Comment (lbergeron) [inline packages/quantic/force-app/main/default/lwc/quanticSearchBox/quanticSearchBox.js:32]: The `@api` annotation is missing for `withoutSubmitButton` otherwise users won't be able to set the attribute.

Same thing for `numberOfSuggestions`. I guess it should be public as well.

Comment (fbeaupreCoveo) [reply packages/quantic/force-app/main/default/lwc/quanticSearchBox/quanticSearchBox.js]: done 👍

**Review by lbergeron — APPROVED** [review]
Looks good to me!

File: packages/quantic/force-app/main/default/lwc/quanticSearchBox/quanticSearchBox.js [hunk 12]
```diff
@@ -80,14 +108,25 @@ export default class QuanticSearchBox extends LightningElement {
...
 
+  setSearchBoxContainerClass() {
+    if(this.withoutSubmitButton){
+      this.searchBoxContainerClass = "slds-combobox__form-element slds-input-has-icon slds-input-has-icon_left-right";
+      this.input.setAttribute("aria-labelledby", "fixed-text-label")
+    } else{
+      this.searchBoxContainerClass = "slds-combobox__form-element slds-input-has-icon slds-input-has-icon_right slds-input-has-fixed-addon";
+      this.input.setAttribute("aria-labelledby", "fixed-text-label fixed-text-addon-post");
+    }
+
```
Comment (nathanlb) [inline packages/quantic/force-app/main/default/lwc/quanticSearchBox/quanticSearchBox.js:119]: extra line here

**Review by nathanlb — APPROVED** [review]
